cirq/work/collector.py

```diff
@@ -0,0 +1,221 @@
+# Copyright 2019 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Optional, Any, Iterable, Union, List, TYPE_CHECKING
+import abc
+import asyncio
+
+import numpy as np
+
+from cirq import circuits, study, value
+from cirq.work import work_pool
+
+if TYPE_CHECKING:
+    # pylint: disable=unused-import
+    import cirq
+
+
+@value.value_equality(unhashable=True)
+class CircuitSampleJob:
+    """Describes a sampling task."""
+
+    def __init__(self,
+                 circuit: circuits.Circuit,
```

Comment:
In cases where the circuit will need to be uploaded, and should only be uploaded once, how will this work?  Will this somehow show up in tag?

Reply:
We'll have to refactor in the future, e.g. by adding a parameters field and expecting the driving method to recognize re-used circuits.